rfcs/proposed/task_group_dynamic_dependencies/implementation_details.md

```diff
@@ -0,0 +1,724 @@
+associated with the task increases the reference counter stored in the dynamic state to prolong it's lifetime until all of the trackers
+are alive.
+
+Empty `tbb::task_tracker` models the empty shared pointer and does not increase or decrease any reference counters.
+
+```cpp
+namespace oneapi {
+namespace tbb {
+
+class task_tracker {
+private:
+    task_dynamic_state* m_state;
+public:
+    task_tracker() : m_state(nullptr) {}
+
+    task_tracker(const task_tracker& other) : m_state(other.m_state) {
+        if (m_state) m_state->reserve();
+    }
+    task_tracker& operator=(const task_tracker& other) { /*same logic*/ }
+
+    task_tracker(task_tracker&& other) : m_state(other.m_state) {
+        other.m_state = nullptr;
+    }
+    task_tracker& operator=(task_tracker&& other) { /*same logic*/ }
+
+    task_tracker(const task_handle& th) : m_state(th ? th.get_dynamic_state() : nullptr) {
+        if (m_state) m_state->reserve();
+    }
+    task_tracker& operator=(const task_handle& th) { /*same logic*/ }
+
+    ~task_tracker() { if (m_state) m_state->release(); }
+};
+
+}
+}
+```
+
+### `task_dynamic_state` in details
+
+As it was mentioned above, `task_dynamic_state` class implements the tracking of task's state (completed or not completed),
+the list of successors and the post-transfer actions. It has the following layout:
+
+```cpp
+class task_dynamic_state {
+private:
+    task_with_dynamic_state*           m_task;
+    std::atomic<successors_list_node*> m_successors_list_head;
+    std::atomic<continuation_vertex*>  m_continuation_vertex;
+    std::atomic<task_dynamic_state*>   m_new_dynamic_state;
+    std::atomic<std::size_t>           m_num_references;
+    small_object_allocator             m_allocator;
+};
+```
+
+`m_task` is a pointer to the task, with which the current dynamic state is associated.
+
+`m_successors_list_head` is an atomic pointer to the head of the successor's list of the currently served task. It is also used
+as a marker of the task completion. It would be described in details in the following sections.
+
+`m_continuation_vertex` is an atomic pointer to eh `continuation_vertex` object associated with the currently served task. The purpose of the
+`continuation_vertex` would be described in the following sections.
+
+`m_new_dynamic_state` is an atomic pointer to the other dynamic state instance. It is used when the currently served task calls
+`transfer_successors_to(new_task)` while executing. `m_new_dynamic_state` will point to `new_task`s dynamic state after doing the transfer.
+It would be described in details in the section about transferring the successors.
+
+`m_num_references` and `m_allocator` are managing the lifetime of the dynamic state.
+
+The stored reference counter is increased when:
+* `task_dynamic_state` associated with the task is created, the reference is reserved for the task object.
+* `task_tracker` instance for the task is created.
```

Comment:
It is legal to have more than one `task_tracker` for a task, I think the added "a" implies that there's not just one `task_tracker` but that you increment the reference count whenever a `task_tracker` is create. Although maybe it should be made more clear that multiple trackers are allowed for the same task.